From 8d21fed1729022bd31adb15014944c8ac2ca9da0 Mon Sep 17 00:00:00 2001
From: Michael Forney <mforney@mforney.org>
Date: Thu, 7 May 2020 00:36:14 -0700
Subject: [PATCH] Revert "Use gnu-case-range and
 gnu-conditional-omitted-operand extensions"

This reverts commit 75a6aa9258270169f43f56e063f1bfb57eebe56b.
---
 Makefile   |  3 +--
 command.c  | 19 ++++++++++---------
 complete.c |  4 ++--
 handle.c   | 45 +++++++++++++++++++++++++++------------------
 ignore.c   |  2 +-
 ui.c       |  7 ++++++-
 6 files changed, 47 insertions(+), 33 deletions(-)

diff --git a/Makefile b/Makefile
index 2c3061d..b844712 100644
--- a/Makefile
+++ b/Makefile
@@ -1,8 +1,7 @@
 PREFIX ?= /usr/local
 MANDIR ?= ${PREFIX}/share/man
 
-CEXTS = gnu-case-range gnu-conditional-omitted-operand
-CFLAGS += -std=c11 -Wall -Wextra -Wpedantic ${CEXTS:%=-Wno-%}
+CFLAGS += -std=c11 -Wall -Wextra -Wpedantic
 LDLIBS = -lncursesw -ltls
 
 -include config.mk
diff --git a/command.c b/command.c
index 4c51433..aec1d4b 100644
--- a/command.c
+++ b/command.c
@@ -69,8 +69,8 @@ static void splitMessage(char *cmd, uint id, char *params) {
 	int overhead = snprintf(
 		NULL, 0, ":%s!%*s@%*s %s %s :\r\n",
 		self.nick,
-		(self.user ? 0 : network.userLen), (self.user ?: "*"),
-		(self.host ? 0 : network.hostLen), (self.host ?: "*"),
+		(self.user ? 0 : network.userLen), (self.user ? self.user : "*"),
+		(self.host ? 0 : network.hostLen), (self.host ? self.host : "*"),
 		cmd, idNames[id]
 	);
 	assert(overhead > 0 && overhead < 512);
@@ -109,7 +109,7 @@ static void commandNotice(uint id, char *params) {
 
 static void commandMe(uint id, char *params) {
 	char buf[512];
-	snprintf(buf, sizeof(buf), "\1ACTION %s\1", (params ?: ""));
+	snprintf(buf, sizeof(buf), "\1ACTION %s\1", (params ? params : ""));
 	echoMessage("PRIVMSG", id, buf);
 }
 
@@ -140,7 +140,7 @@ static void commandPart(uint id, char *params) {
 
 static void commandQuit(uint id, char *params) {
 	(void)id;
-	set(&self.quit, (params ?: "nyaa~"));
+	set(&self.quit, (params ? params : "nyaa~"));
 }
 
 static void commandNick(uint id, char *params) {
@@ -226,7 +226,7 @@ static void commandOp(uint id, char *params) {
 }
 
 static void commandDeop(uint id, char *params) {
-	channelListMode(id, '-', 'o', (params ?: self.nick));
+	channelListMode(id, '-', 'o', (params ? params : self.nick));
 }
 
 static void commandVoice(uint id, char *params) {
@@ -238,7 +238,7 @@ static void commandVoice(uint id, char *params) {
 }
 
 static void commandDevoice(uint id, char *params) {
-	channelListMode(id, '-', 'v', (params ?: self.nick));
+	channelListMode(id, '-', 'v', (params ? params : self.nick));
 }
 
 static void commandBan(uint id, char *params) {
@@ -399,11 +399,12 @@ static void commandExec(uint id, char *params) {
 	if (pid < 0) err(EX_OSERR, "fork");
 	if (pid) return;
 
+	const char *shell = getenv("SHELL");
+	if (!shell) shell = "/bin/sh";
+
 	close(STDIN_FILENO);
 	dup2(execPipe[1], STDOUT_FILENO);
 	dup2(utilPipe[1], STDERR_FILENO);
-
-	const char *shell = getenv("SHELL") ?: "/bin/sh";
 	execlp(shell, shell, "-c", params, NULL);
 	warn("%s", shell);
 	_exit(EX_UNAVAILABLE);
@@ -418,7 +419,7 @@ static void commandHelp(uint id, char *params) {
 	if (pid) return;
 
 	char buf[256];
-	snprintf(buf, sizeof(buf), "ip%s$", (params ?: "COMMANDS"));
+	snprintf(buf, sizeof(buf), "ip%s$", (params ? params : "COMMANDS"));
 	setenv("LESS", buf, 1);
 	execlp("man", "man", "1", "catgirl", NULL);
 	dup2(utilPipe[1], STDERR_FILENO);
diff --git a/complete.c b/complete.c
index 5835926..0e1de47 100644
--- a/complete.c
+++ b/complete.c
@@ -71,7 +71,7 @@ static struct Node *prepend(struct Node *node) {
 	node->next = head;
 	if (head) head->prev = node;
 	head = node;
-	tail = (tail ?: node);
+	if (!tail) tail = node;
 	return node;
 }
 
@@ -80,7 +80,7 @@ static struct Node *append(struct Node *node) {
 	node->prev = tail;
 	if (tail) tail->next = node;
 	tail = node;
-	head = (head ?: node);
+	if (!head) head = node;
 	return node;
 }
 
diff --git a/handle.c b/handle.c
index e5e370e..04b73d3 100644
--- a/handle.c
+++ b/handle.c
@@ -280,9 +280,9 @@ static void handleReplyISupport(struct Message *msg) {
 			set(&network.setParamModes, setParam);
 			set(&network.channelModes, channel);
 		} else if (!strcmp(key, "EXCEPTS")) {
-			network.excepts = (msg->params[i] ?: "e")[0];
+			network.excepts = (msg->params[i] ? msg->params[i][0] : 'e');
 		} else if (!strcmp(key, "INVEX")) {
-			network.invex = (msg->params[i] ?: "I")[0];
+			network.invex = (msg->params[i] ? msg->params[i][0] : 'I');
 		}
 	}
 }
@@ -329,7 +329,7 @@ static void handleJoin(struct Message *msg) {
 		"\3%02d%s\3\t%s%s%sarrives in \3%02d%s\3",
 		hash(msg->user), msg->nick,
 		(msg->params[2] ? "(" : ""),
-		(msg->params[2] ?: ""),
+		(msg->params[2] ? msg->params[2] : ""),
 		(msg->params[2] ? ") " : ""),
 		hash(msg->params[0]), msg->params[0]
 	);
@@ -361,12 +361,14 @@ static void handlePart(struct Message *msg) {
 		id, heat, tagTime(msg),
 		"\3%02d%s\3\tleaves \3%02d%s\3%s%s",
 		hash(msg->user), msg->nick, hash(msg->params[0]), msg->params[0],
-		(msg->params[1] ? ": " : ""), (msg->params[1] ?: "")
+		(msg->params[1] ? ": " : ""),
+		(msg->params[1] ? msg->params[1] : "")
 	);
 	logFormat(
 		id, tagTime(msg), "%s leaves %s%s%s",
 		msg->nick, msg->params[0],
-		(msg->params[1] ? ": " : ""), (msg->params[1] ?: "")
+		(msg->params[1] ? ": " : ""),
+		(msg->params[1] ? msg->params[1] : "")
 	);
 }
 
@@ -383,12 +385,14 @@ static void handleKick(struct Message *msg) {
 		hash(msg->user), msg->nick,
 		completeColor(id, msg->params[1]), msg->params[1],
 		hash(msg->params[0]), msg->params[0],
-		(msg->params[2] ? ": " : ""), (msg->params[2] ?: "")
+		(msg->params[2] ? ": " : ""),
+		(msg->params[2] ? msg->params[2] : "")
 	);
 	logFormat(
 		id, tagTime(msg), "%s kicks %s out of %s%s%s",
 		msg->nick, msg->params[1], msg->params[0],
-		(msg->params[2] ? ": " : ""), (msg->params[2] ?: "")
+		(msg->params[2] ? ": " : ""),
+		(msg->params[2] ? msg->params[2] : "")
 	);
 	completeRemove(id, msg->params[1]);
 	if (kicked) completeClear(id);
@@ -427,13 +431,15 @@ static void handleQuit(struct Message *msg) {
 			id, heat, tagTime(msg),
 			"\3%02d%s\3\tleaves%s%s",
 			hash(msg->user), msg->nick,
-			(msg->params[0] ? ": " : ""), (msg->params[0] ?: "")
+			(msg->params[0] ? ": " : ""),
+			(msg->params[0] ? msg->params[0] : "")
 		);
 		if (id == Network) continue;
 		logFormat(
 			id, tagTime(msg), "%s leaves%s%s",
 			msg->nick,
-			(msg->params[0] ? ": " : ""), (msg->params[0] ?: "")
+			(msg->params[0] ? ": " : ""),
+			(msg->params[0] ? msg->params[0] : "")
 		);
 	}
 	completeRemove(None, msg->nick);
@@ -605,7 +611,7 @@ static void handleReplyUserModeIs(struct Message *msg) {
 	for (char *ch = msg->params[1]; *ch; ++ch) {
 		if (*ch == '+') continue;
 		const char *name = UserModes[(byte)*ch];
-		catf(&cat, ", +%c%s%s", *ch, (name ? " " : ""), (name ?: ""));
+		catf(&cat, ", +%c%s%s", *ch, (name ? " " : ""), (name ? name : ""));
 	}
 	uiFormat(
 		Network, Warm, tagTime(msg),
@@ -647,13 +653,13 @@ static void handleReplyChannelModeIs(struct Message *msg) {
 			assert(param < ParamCap);
 			catf(
 				&cat, ", +%c%s%s %s",
-				*ch, (name ? " " : ""), (name ?: ""),
+				*ch, (name ? " " : ""), (name ? name : ""),
 				msg->params[param++]
 			);
 		} else {
 			catf(
 				&cat, ", +%c%s%s",
-				*ch, (name ? " " : ""), (name ?: "")
+				*ch, (name ? " " : ""), (name ? name : "")
 			);
 		}
 	}
@@ -680,7 +686,7 @@ static void handleMode(struct Message *msg) {
 				hash(msg->user), msg->nick,
 				(set ? "" : "un"),
 				self.color, msg->params[0],
-				set["-+"], *ch, (name ? " " : ""), (name ?: "")
+				set["-+"], *ch, (name ? " " : ""), (name ? name : "")
 			);
 		}
 		return;
@@ -837,7 +843,7 @@ static void handleErrorBanListFull(struct Message *msg) {
 	require(msg, false, 4);
 	uiFormat(
 		idFor(msg->params[1]), Cold, tagTime(msg),
-		"%s", (msg->params[4] ?: msg->params[3])
+		"%s", (msg->params[4] ? msg->params[4] : msg->params[3])
 	);
 }
 
@@ -970,14 +976,15 @@ static void handleReplyWhoisIdle(struct Message *msg) {
 		}
 	}
 	char signon[sizeof("0000-00-00 00:00:00")];
-	time_t time = strtol((msg->params[3] ?: ""), NULL, 10);
+	time_t time = (msg->params[3] ? strtol(msg->params[3], NULL, 10) : 0);
 	strftime(signon, sizeof(signon), "%F %T", localtime(&time));
 	uiFormat(
 		Network, Warm, tagTime(msg),
 		"\3%02d%s\3\tis idle for %lu %s%s%s%s",
 		completeColor(Network, msg->params[1]), msg->params[1],
 		idle, unit, (idle != 1 ? "s" : ""),
-		(msg->params[3] ? ", signed on " : ""), (msg->params[3] ? signon : "")
+		(msg->params[3] ? ", signed on " : ""),
+		(msg->params[3] ? signon : "")
 	);
 }
 
@@ -1010,7 +1017,9 @@ static void handleReplyWhoisGeneric(struct Message *msg) {
 		Network, Warm, tagTime(msg),
 		"\3%02d%s\3\t%s%s%s",
 		completeColor(Network, msg->params[1]), msg->params[1],
-		msg->params[2], (msg->params[3] ? " " : ""), (msg->params[3] ?: "")
+		msg->params[2],
+		(msg->params[3] ? " " : ""),
+		(msg->params[3] ? msg->params[3] : "")
 	);
 }
 
@@ -1063,7 +1072,7 @@ static bool isMention(const struct Message *msg) {
 	const char *match = msg->params[1];
 	while (NULL != (match = strstr(match, self.nick))) {
 		char a = (match > msg->params[1] ? match[-1] : ' ');
-		char b = (match[len] ?: ' ');
+		char b = (match[len] ? match[len] : ' ');
 		if ((isspace(a) || ispunct(a)) && (isspace(b) || ispunct(b))) {
 			return true;
 		}
diff --git a/ignore.c b/ignore.c
index f8e4d59..73d2d90 100644
--- a/ignore.c
+++ b/ignore.c
@@ -79,7 +79,7 @@ enum Heat ignoreCheck(enum Heat heat, uint id, const struct Message *msg) {
 	snprintf(
 		match, sizeof(match), "%s!%s@%s %s %s %s",
 		msg->nick, msg->user, msg->host,
-		msg->cmd, idNames[id], (msg->params[1] ?: "")
+		msg->cmd, idNames[id], msg->params[1] ? msg->params[1] : ""
 	);
 	for (size_t i = 0; i < ignore.len; ++i) {
 		if (fnmatch(ignore.patterns[i], match, FNM_CASEFOLD)) continue;
diff --git a/ui.c b/ui.c
index 22ec02e..c01da59 100644
--- a/ui.c
+++ b/ui.c
@@ -836,7 +836,6 @@ static void keyCode(int code) {
 		break; case KeyMetaGt: windowScroll(window, -BufferCap);
 		break; case KeyMetaLt: windowScroll(window, +BufferCap);
 
-		break; case KeyMeta0 ... KeyMeta9: uiShowNum(code - KeyMeta0);
 		break; case KeyMetaA: showAuto();
 		break; case KeyMetaB: edit(id, EditPrevWord, 0);
 		break; case KeyMetaD: edit(id, EditDeleteNextWord, 0);
@@ -862,6 +861,12 @@ static void keyCode(int code) {
 		break; case KEY_SEND: windowScroll(window, -BufferCap);
 		break; case KEY_SHOME: windowScroll(window, +BufferCap);
 		break; case KEY_UP: windowScroll(window, +1);
+
+		break; default: {
+			if (code >= KeyMeta0 && code <= KeyMeta9) {
+				uiShowNum(code - KeyMeta0);
+			}
+		}
 	}
 }
 
-- 
2.28.0

